### PR TITLE
Add lbinomial function based on lbeta

### PR DESCRIPTION
**Pull request by freeboson**
This was originally submitted to base in https://github.com/JuliaLang/julia/pull/27419, but https://github.com/JuliaLang/julia/pull/27473 moved `special/gamma.jl` things here, so I'm making the same PR here now. Details are copied over:

This implementation of `lbinomial` uses the fact that `binomial(n,k) = 1/( (n+1) * B(n - k + 1, k + 1) )`. I have added tests for some simple cases, and then also comparing for very large `n` with `binomial(::BigInt, ::BigInt)`.

The main motivation is evaluating `log(binomial(n,k))` for large `n` with `k` near `n/2` without approximating with Poisson:

```julia
julia> log(binomial(70, 25)) # OK
43.311504703669215

julia> log(binomial(70, 26)) # :(
ERROR: InexactError()
Stacktrace:
 [1] binomial(::Int64, ::Int64) at ./intfuncs.jl:810

julia> log(binomial(BigInt(70), BigInt(26))) # OK, if you can afford it
4.386007065541805521142055198415251198714576775237582336220765505130378175238536e+01

julia> ans - lbinomial(70, 26) # Close enough?
-8.056495743460440321531836165841374176637792344948696218247614638531207882439305e-15
```

In R, this function is available as `lchoose(n, k)` and is a builtin.

### CODE REVIEW

File: src/gamma.jl [hunk 1]
```diff
@@ -764,3 +764,27 @@ const lfactorial = lfact
...
+    lbinomial(n, k) = log(abs(binomial(n, k)))
+
+Accurate natural logarithm of the absolute value of the [`binomial`](@ref)
+coefficient `binomial(n, k)` for large `n` and `k` near `n/2`.
+"""
+function lbinomial(n::T, k::T) where {T<:Integer}
+    S = float(T)
+    (k < 0) && return typemin(S)
+    if n < 0
+       n = -n + k - 1
```
Comment (ararslan) [inline src/gamma.jl]: Incorrect indentation here (needs one more space)

File: src/gamma.jl [hunk 2]
```diff
@@ -764,3 +764,27 @@ const lfactorial = lfact
...
+    end
+    k > n && return typemin(S)
+    (k == 0 || k == n) && return zero(S)
+    (k == 1) && return log(abs(n))
+    if k > (n>>1)
+       k = n - k
```
Comment (ararslan) [inline src/gamma.jl]: Here too

File: src/gamma.jl [hunk 3]
```diff
@@ -764,3 +764,27 @@ const lfactorial = lfact
...
+    end
+    -log1p(n) - lbeta(n - k + one(T), k + one(T))
+end
+lbinomial(n::Integer, k::Integer) = lbinomial(promote(n, k)...)
+
+export lbinomial
```
Comment (ararslan) [inline src/gamma.jl]: This should go with the other exports.

File: test/runtests.jl [hunk 4]
```diff
@@ -636,3 +636,17 @@ end
 
     @test beta(big(1.0),big(1.2)) ≈ beta(1.0,1.2) rtol=4*eps()
 end
+
+@testset "lbinomial" begin
+    lbfixed(n) = k -> lbinomial(n, k)
```
Comment (ararslan) [inline test/runtests.jl]: Why is this necessary? You can write `lbfixed(200).(0:200)` below as `lbinomial.(200, 0:200)`.

File: src/gamma.jl (hunk 3 above)
Comment (freeboson) [reply src/gamma.jl]: I put it now with `cosint`. Not sure if this is the "normal" place or not.

File: test/runtests.jl (hunk 4 above)
Comment (ararslan) [reply test/runtests.jl]: Note that the same applies to the function defined directly below this line.

**Comment by freeboson** [comment]
Rebased with the corrections, dropped the lambdas in testing.

**Comment by simonbyrne** [comment]
Thanks!

**Comment by ararslan** [comment]
It's deprecated in favor of `logabsbinomial`:
https://github.com/JuliaMath/SpecialFunctions.jl/blob/8136181a2f7b44d90f8f486e27c013cb747ceb32/src/deprecated.jl#L14
